waiter/src/waiter/reporter.clj

```diff
@@ -0,0 +1,151 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns waiter.reporter
+  (:require [clj-time.core :as t]
+            [metrics.core :as metrics]
+            [schema.core :as s]
+            [waiter.schema :as schema]
+            [clojure.tools.logging :as log])
+  (:import (com.codahale.metrics ConsoleReporter MetricFilter ScheduledReporter)
+           (com.codahale.metrics.graphite Graphite GraphiteReporter GraphiteSender PickledGraphite)
+           java.io.PrintStream
+           java.net.InetSocketAddress
+           java.util.concurrent.TimeUnit))
+
+(defprotocol CodahaleReporter
+  "A reporter for codahale metrics"
+
+  (close! [this]
+    "Stops the reporter and performs any cleanup")
+
+  (state [this]
+    "Returns the state of this reporter"))
+
+(defn- make-metrics-filter
+  [filter-regex]
+  (reify MetricFilter (matches [_ name _] (some? (re-matches filter-regex name)))))
+
+(defn- make-codahale-reporter
+  [^ScheduledReporter scheduled-reporter state period-ms]
+  (.start scheduled-reporter period-ms TimeUnit/MILLISECONDS)
+  (reify CodahaleReporter
+    (close! [_] (.close scheduled-reporter) (reset! state :closed))
+    (state [_] @state)))
+
+
+(defn validate-console-reporter-config
+  "Validates ConsoleReporter settings and sets defaults"
+  [config]
+  (s/validate {(s/required-key :period-ms) schema/positive-int
+               (s/optional-key :filter-regex) s/Regex
+               s/Any s/Any} config))
+
+(defn make-console-reporter
+  "Creates a ConsoleReporter for codahale metrics"
+  ([filter-regex] (make-console-reporter filter-regex nil))
+  ([filter-regex ^PrintStream output]
+   (let [state (atom {})
+         console-reporter (cond-> (ConsoleReporter/forRegistry metrics/default-registry)
+                                  output (.outputTo output)
+                                  filter-regex (.filter (make-metrics-filter filter-regex))
+                                  true (-> (.convertRatesTo TimeUnit/SECONDS)
+                                           (.convertDurationsTo TimeUnit/MILLISECONDS)
+                                           (.build)))]
+     [console-reporter state])))
```

Comment:
Why not just return the `console-reporter`. It provides a `state` function to retrieve the state.

Reply:
No, it doesn't. Console reporter is from the library we're using, state is my atom. Then we combine them when we make the CodahaleReporter protocol instance

Reply:
Okay.